snakemake/executors.py

```diff
@@ -2168,3 +2171,878 @@ def run_wrapper(
+            request = operations.cancel(name=job.jobname)
+            logger.debug("Cancelling operation {}".format(job.jobid))
+            try:
+                self._retry_request(request)
+            except (Exception, BaseException, googleapiclient.errors.HttpError):
+                continue
+
+        self.shutdown()
+
+    def get_available_machine_types(self):
+        """Using the regions available at self.regions, use the GCP API
+           to retrieve a lookup dictionary of all available machine types.
+        """
+        # Regular expression to determine if zone in region
+        regexp = "^(%s)" % "|".join(self.regions)
+
+        # Retrieve zones, filter down to selected regions
+        zones = self._retry_request(
+            self._compute_cli.zones().list(project=self.project)
+        )
+        zones = [z for z in zones["items"] if re.search(regexp, z["name"])]
+
+        # Retrieve machine types available across zones
+        # https://cloud.google.com/compute/docs/regions-zones/
+        lookup = {}
+        for zone in zones:
+            request = self._compute_cli.machineTypes().list(
+                project=self.project, zone=zone["name"]
+            )
+            lookup[zone["name"]] = self._retry_request(request)["items"]
+
+        # Only keep those that are shared, use last zone as a base
+        machine_types = {mt["name"]: mt for mt in lookup[zone["name"]]}
+        del lookup[zone["name"]]
+
+        # Update final list based on the remaining
+        to_remove = set()
+        for zone, types in lookup.items():
+            names = [x["name"] for x in types]
+            names = [name for name in names if "micro" not in name]
+            names = [name for name in names if not re.search("^(e2|m1)", name)]
+            for machine_type in list(machine_types.keys()):
+                if machine_type not in names:
+                    to_remove.add(machine_type)
+
+        for machine_type in to_remove:
+            del machine_types[machine_type]
+        return machine_types
+
+    def _add_gpu(self, gpu_count):
+        """Add a number of NVIDIA gpus to the current executor. This works
+           by way of adding nvidia_gpu to the job default resources, and also
+           changing the default machine type prefix to be n1, which is
+           the currently only supported instance type for using GPUs for LHS.
+        """
+        if not gpu_count or gpu_count == 0:
+            return
+
+        logger.debug(
+            "found resource request for {} GPUs. This will limit to n1 "
+            "instance types.".format(gpu_count)
+        )
+        self.workflow.default_resources.parsed["nvidia_gpu"] = gpu_count
+        self.workflow.default_resources.args.append("nvidia_gpu=%s" % gpu_count)
+        self._machine_type_prefix = self._machine_type_prefix or ""
+        if not self._machine_type_prefix.startswith("n1"):
+            self._machine_type_prefix = "n1"
```

Comment:
Why does this set an executor attribute? It is a method called while handling a single job, right?

Reply:
We need the machine type prefix in other places for the executor, and it's an easy / consistent way to do it instead of passing around variables. If a gpu is chosen it must be in the n1 family.